### ISSUE TITLE
parser failing on valid query with parameters?

### ISSUE THREAD

**Issue by Fhffsthu** [angel-dart/graphql#23]
Some test code:

   ```dart
 final List<Token> tokens = scan(meQuery);
    final Parser parser = Parser(tokens);

    if (parser.errors.isNotEmpty) {
      // Handle errors...
      print(parser.errors.toString());
    }

    // Parse the GraphQL document using recursive descent
    final DocumentContext doc = parser.parseDocument();

    expect(doc.definitions, isNotNull);
    expect(doc.definitions, isNotEmpty);

```
Fails with this query:

```graphql
query customerMemberAttributes($memberId: Int!){
  customerByCustomerId(customerId: 7) {
    customerMemberAttributesByCustomerId {
      nodes {
        customerMemberAttributeId
        memberAttributesByCustomerMemberAttributeId(condition: {memberId: $memberId}) {
          nodes {
            memberAttributeId
          }
        }
      }
    }
  }
}
```
but not this query:

```graphql
query customerMemberAttributes($memberId: Int!){
  customerByCustomerId(customerId: 7) {
    customerMemberAttributesByCustomerId {
      nodes {
        customerMemberAttributeId
        memberAttributesByCustomerMemberAttributeId(condition: {memberId: 7}) {
          nodes {
            memberAttributeId
          }
        }
      }
    }
  }
}
```

I believe these queries are valid and both should work.

**Comment by thosakwe** [comment]
Please format the second query (can't read it).

**Comment by Fhffsthu** [comment]
Sorry about the formatting - fixed.

**Comment by thosakwe** [comment]
Okay, I need more information. The word "fails"/"failing" really doesn't me a clue as to what is happening.

1. I notice backslashes in front of your `$`. Is there a specific reason for this?
2. Are you getting an error message, etc. when parsing the second query? If so, please paste its contents, *as well as the associated source span*.
3. What are you expecting to happen here? What is actually happening?

To the get the span:

```dart
for (var e in errors) {
  // Do whatever other printing is necessary...
  print(e.message);
  print(e.span.highlight());
}
```

**Comment by Fhffsthu** [comment]
Sorry about the backslashes - needed to escape the "$".
The first query is what is actually failing with these errors:
```Missing value after ":".
  ╷
1 │ query customerMemberAttributes($memberId: Int!){   customerByCustomerId(customerId: 7) {     customerMemberAttributesByCustomerId {       nodes {         customerMemberAttributeId         memberAttributesByCustomerMemberAttributeId(condition: {memberId: $memberId}) {           nodes {             memberAttributeId           }         }       }     }   } } 
  │                                                                                                                                                                                                                                                             ^
  ╵
Unterminated object literal.
  ╷
1 │ query customerMemberAttributes($memberId: Int!){   customerByCustomerId(customerId: 7) {     customerMemberAttributesByCustomerId {       nodes {         customerMemberAttributeId         memberAttributesByCustomerMemberAttributeId(condition: {memberId: $memberId}) {           nodes {             memberAttributeId           }         }       }     }   } } 
  │                                                                                                                                                                                                                                                    ^
  ╵
Missing ")" in argument list.
  ╷
1 │ query customerMemberAttributes($memberId: Int!){   customerByCustomerId(customerId: 7) {     customerMemberAttributesByCustomerId {       nodes {         customerMemberAttributeId         memberAttributesByCustomerMemberAttributeId(condition: {memberId: $memberId}) {           nodes {             memberAttributeId           }         }       }     }   } } 
  │                                                                                                                                                                                                                                        ^
  ╵
Missing "}" after selection set.
  ╷
1 │ query customerMemberAttributes($memberId: Int!){   customerByCustomerId(customerId: 7) {     customerMemberAttributesByCustomerId {       nodes {         customerMemberAttributeId         memberAttributesByCustomerMemberAttributeId(condition: {memberId: $memberId}) {           nodes {             memberAttributeId           }         }       }     }   } } 
  │                                                                                                                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ╵
Missing "}" after selection set.
  ╷
1 │ query customerMemberAttributes($memberId: Int!){   customerByCustomerId(customerId: 7) {     customerMemberAttributesByCustomerId {       nodes {         customerMemberAttributeId         memberAttributesByCustomerMemberAttributeId(condition: {memberId: $memberId}) {           nodes {             memberAttributeId           }         }       }     }   } } 
  │                                                                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ╵
Missing "}" after selection set.
  ╷
1 │ query customerMemberAttributes($memberId: Int!){   customerByCustomerId(customerId: 7) {     customerMemberAttributesByCustomerId {       nodes {         customerMemberAttributeId         memberAttributesByCustomerMemberAttributeId(condition: {memberId: $memberId}) {           nodes {             memberAttributeId           }         }       }     }   } } 
  │                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ╵
Missing selection set in fragment definition.
  ╷
1 │ query customerMemberAttributes($memberId: Int!){   customerByCustomerId(customerId: 7) {     customerMemberAttributesByCustomerId {       nodes {         customerMemberAttributeId         memberAttributesByCustomerMemberAttributeId(condition: {memberId: $memberId}) {           nodes {             memberAttributeId           }         }       }     }   } } 
  │       ^^^^^^^^^^^^^^^^^^^^^^^^
  ╵
```

The definitions should not be empty (I believe).
I am trying to get the "operation name" (which is the definition.name) for the query. The second query returns the name and the first does not. I actually need to pass the variable into the query in the code so hard coding the memberId is a no go.
As a note I am digging into a flutter GraphQL library issue and have run it down to this library which is used by them.

**Comment by thosakwe** [comment]
Ah, I see.

So, peering back into the spec (the original parser is from 2016), it looks like `Variable` nodes can now be considered `Value` nodes: https://graphql.github.io/graphql-spec/June2018/#Value

However, `ObjectFieldContext` only takes scalar values:
https://github.com/angel-dart/graphql/blob/master/graphql_parser/lib/src/language/ast/misc_value.dart#L65

The fix necessary will be to be to ultimately allow object fields to take `ValueOrVariable` instead of just `Value`. This will be a much bigger change than it might seem.

The steps necessary to prevent this change from breaking *all* of the other GraphQL packages are:
* Add a `T computeValue(Map<String, dynamic> variables)` to `ValueContext`.
  * This should default to returning `ValueContext.value`.
  * Add logic for resolving variable names in object fields from the `variables` argument.
  * Throw an error if the plain getter `ObjectValueContext.value` is called when any field contains a variable.
* Modify `graphql_server to call `computeValue` instead of just getting `value`.

**Comment by thosakwe** [comment]
A fix for this is rolling in branch `1.2.0`: https://github.com/angel-dart/graphql/tree/1.2.0
Working on getting this release out ASAP. I aim to fix all outstanding bugs in this package.

**Comment by thosakwe** [comment]
Will go ahead and close this, as it's been solved (though it has not been released yet).
